src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs

```diff
@@ -177,23 +184,36 @@ protected void AssertSelf ()
 			if (!request.RequestUri.IsAbsoluteUri)
 				throw new ArgumentException ("Must represent an absolute URI", "request");
 
-			URL java_url = new URL (request.RequestUri.ToString ());
-			URLConnection java_connection = java_url.OpenConnection ();
-			HttpURLConnection httpConnection = await SetupRequestInternal (request, java_connection);
-			return await ProcessRequest (request, java_url, httpConnection, cancellationToken);
+			var redirectState = new RequestRedirectionState {
+				NewUrl = request.RequestUri,
+				RedirectCounter = 0,
+				Method = request.Method
+			};
+			while (true) {
+				URL java_url = new URL (redirectState.NewUrl.ToString ());
+				URLConnection java_connection = java_url.OpenConnection ();
+				HttpURLConnection httpConnection = await SetupRequestInternal (request, java_connection);
+				HttpResponseMessage response = await ProcessRequest (request, java_url, httpConnection, cancellationToken, redirectState);
+				if (response != null)
+					return response;
+
+				if (redirectState.NewUrl == null)
+					throw new InvalidOperationException ("Request redirected but no new URI specified");
+				request.Method = redirectState.Method;
+			}
 		}
 		
-		Task <HttpResponseMessage> ProcessRequest (HttpRequestMessage request, URL javaUrl, HttpURLConnection httpConnection, CancellationToken cancellationToken)
+		Task <HttpResponseMessage> ProcessRequest (HttpRequestMessage request, URL javaUrl, HttpURLConnection httpConnection, CancellationToken cancellationToken, RequestRedirectionState redirectState)
 		{
 			cancellationToken.ThrowIfCancellationRequested ();
-			httpConnection.InstanceFollowRedirects = AllowAutoRedirect;
```

Comment:
Does this imply that we need to update the docs on the `AllowAutoRedirect` property?